openedx/core/djangoapps/content/block_structure/models.py

```diff
@@ -59,6 +61,63 @@ def _bs_model_storage():
     )
 
 
+class CustomizableFileField(models.FileField):
+    """
+    Subclass of FileField that allows custom settings to not
+    be serialized (hard-coded) in migrations. Otherwise,
+    migrations include optional settings for storage (such as
+    the storage class and bucket name); we don't want to
+    create new migration files for each configuration change.
+    """
+    def __init__(self, *args, **kwargs):
+        kwargs.update(dict(
+            upload_to=_path_name,
+            storage=_bs_model_storage(),
+            max_length=500,  # allocate enough for base path + prefix + usage_key + timestamp in filepath
+        ))
+        super(CustomizableFileField, self).__init__(*args, **kwargs)
+
+    def deconstruct(self):
+        name, path, args, kwargs = super(CustomizableFileField, self).deconstruct()
+        del kwargs['upload_to']
+        del kwargs['storage']
+        del kwargs['max_length']
+        return name, path, args, kwargs
+
+
+@contextmanager
+def _storage_error_handling(bs_model, operation, is_read_operation=False):
```

Comment:
Some quick impressions:

1. Retries are completely reasonable, but yeah, at some point we need to say "this didn't work, try again".
2. I don't believe we have this code for S3 specifically, but I'm assuming it's just a for loop with longer timeout intervals + jitter? We could also delay the celery retries by some function of the attempt number, if we're not doing so already.
3. Short term: The first priority would seem to be to never leave the block structures data model in a broken/inconsistent state. But as long as we have a confirmed write to S3 before we change the pointer on the database model, we should be good there.
4. We have a management command of some sort to kick these things off across all courses, for error recovery, right? As long as we have that, I don't think we need to be able to automatically recover from an S3-apocalypse in the near term -- or rather, that we shouldn't invest too heavily in that aspect just yet.
5. Long term: We've gotten away with a certain amount of drift in publishing time because it usually doesn't take very long. But between failure modes and more complex publishing tasks (and eventually backing field data with block structures), we'll eventually want a mechanism where the LMS says definitively, "this is the current published version on the LMS" and Studio can check that to see if it matches what's in its modulestore -- particularly if we're serious about breaking up Studio and the LMS at some point. Right now it's a hodgepodge of "listen for the publish signal, and do your thing". There's a lot assumed in that worldview though, so it might be a bit off-topic for this particular issue.